source/common/config/grpc_mux_impl.cc

```diff
@@ -40,40 +29,13 @@ GrpcMuxImpl::~GrpcMuxImpl() {
 
 void GrpcMuxImpl::start() { establishNewStream(); }
 
-void GrpcMuxImpl::setRetryTimer() {
-  retry_timer_->enableTimer(std::chrono::milliseconds(backoff_strategy_->nextBackOffMs()));
-}
-
-void GrpcMuxImpl::establishNewStream() {
-  ENVOY_LOG(debug, "Establishing new gRPC bidi stream for {}", service_method_.DebugString());
-  stream_ = async_client_->start(service_method_, *this);
-  if (stream_ == nullptr) {
-    ENVOY_LOG(warn, "Unable to establish new stream");
-    handleFailure();
-    return;
-  }
-
-  control_plane_stats_.connected_state_.set(1);
-  for (const auto type_url : subscriptions_) {
-    queueDiscoveryRequest(type_url);
-  }
-}
-
 void GrpcMuxImpl::drainRequests() {
   ENVOY_LOG(trace, "draining discovery requests {}", request_queue_.size());
-  while (!request_queue_.empty()) {
+  while (!request_queue_.empty() && checkRateLimitAllowsDrain(request_queue_.size())) {
```

Comment:
Why couldn't this whole method move to the superclass? I'd prefer to avoid changing behavior during the simple refactor  and then do a separate PR for any goodness needed to support the on-demand case (with tests as needed).

Reply:
It can, and so could queueDiscoveryRequest; DiscoveryGrpcStream just needs to be templated on queue item type as well. I will do that.

Keeping the fix in a separate PR sounds good. However, I think you misunderstood me: I am not suggesting this change to support new things, rather, I think the logic is incorrect even for the existing vanilla xDS case. DiscoveryRequests can be silently dropped when the gRPC channel briefly goes down. That seems wrong to me, although I wanted to double-check that it isn't intended.

Reply:
I don't think this is an issue today; when the stream is re-established, a new DiscoverRequest will be sent, so any requests lost while disconnected don't matter.